libsrc/_DEVELOPMENT/target/rc2014/startup/rc2014_crt_128.asm.m4

```diff
@@ -0,0 +1,397 @@
+
+IF (__crt_enable_commandline = 1) || (__crt_enable_commandline >= 3)
+
+    include "../crt_cmdline_empty.inc"
+
+ENDIF
+
+.__Restart_2
+
+IF __crt_enable_commandline >= 1
+
+    push hl                     ; argv
+    push bc                     ; argc
+
+ENDIF
+
+SECTION code_crt_init          ; user and library initialization
+
+; The classic CRTs need some things setup, so do it
+
+    SECTION code_crt_init
+
+IF ( __crt_model = 1 )
+
+    ; rom model + data section is not compressed
+
+    EXTERN  __CODE_END_tail
+    EXTERN  __DATA_head, __DATA_END_tail
+
+    ld hl,__CODE_END_tail
+    ld de,__DATA_head
+    ld bc,__DATA_END_tail-__DATA_head
+
+    EXTERN  asm_memcpy
+    call asm_memcpy
+
+ELIF ( __crt_model >= 2 )
+
+    ; rom model + data section is compressed
+
+    EXTERN  __CODE_END_tail
+    EXTERN  __DATA_head, __DATA_END_tail
+
+    ld hl,__CODE_END_tail
+    ld de,__DATA_head
+    ld bc,__DATA_END_tail-__DATA_head
+
+    ld a,b
+    or c
+
+  IF ( __crt_model = 2)
+    EXTERN  asm_dzx7_standard
+    call NZ, asm_dzx7_standard
+  ELIF ( __crt_model = 3)
+    EXTERN  asm_dzx0_standard
+    call NZ, asm_dzx0_standard
+  ENDIF
+
+ENDIF
+
+IF __crt_initialize_bss = 1
+
+    EXTERN  __BSS_head
+    EXTERN  __BSS_UNINITIALIZED_head
+
+    ld hl,__BSS_head
+    ld bc,__BSS_UNINITIALIZED_head-__BSS_head-1
+
+.init_bss_8085
+    ld (hl),0
+    inc hl
+    dec bc
+    jp NK,init_bss_8085
+
+ENDIF
+
+IF CRT_ENABLE_STDIO = 1
+
+    ; Setup std* streams
+    ld hl,__sgoioblk+2
+    ld (hl),19                  ; stdin
+    ld hl,__sgoioblk+12
+    ld (hl),21                  ; stdout
+    ld hl,__sgoioblk+22
+    ld (hl),21                  ; stderr
+
+ENDIF
+
+    ; copy interrupt vector table to final location
+
+    include "../crt_set_interrupt_mode.inc"
+
+    include "../../../../lib/crt/classic/crt_init_heap.inc"
+
+
+SECTION code_crt_init           ; user and library initialization
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; MAIN ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+SECTION code_crt_main
+
+    include "../crt_start_ei.inc"
+
+    ; call user program
+
+    call _main                  ; hl = return status
+
+    ; run exit stack
+
+IF __clib_exit_stack_size > 0
+
+    EXTERN asm_exit
+    jp asm_exit                 ; exit function jumps to __Exit
+
+ENDIF
+
+.__Exit
+
+IF !((__crt_on_exit & 0x10000) && (__crt_on_exit & 0x8))
+
+    ; not restarting
+
+    push hl                     ; save return status
+
+ENDIF
+
+SECTION code_crt_exit           ; user and library cleanup
+SECTION code_crt_return
+
+    ; close files
+
+    include "../clib_close.inc"
+
+    ; terminate
+
+IF (__crt_on_exit = 0x10002)
+
+    ; returning to basic
+
+    pop hl
+
+    ex de,hl
+    ld hl,(__sp_or_ret)
+    ld sp,hl
+    ex de,hl
+
+    IF CRT_ABPASS > 0
+
+        ld a,h
+        ld b,l
+        jp CRT_ABPASS
+
+    ENDIF
+
+    ei
```

Comment:
If CRT_ABPASS > 0, shouldn't EI instruction be execute too?

Reply:
Thanks for the input.

I'll have a look at this when I'm back on it next week (afk currently).

But there's still quite a bit of work to do here, including reinstating the cpm subtype (required to access FatFS drives on the rc2014. Doh!).

Once I build / test the hardware then I can write and test the actual drivers too.

Reply:
I've moved the `ei` before a potential jump to the BASIC `ABPASS` function.
There would be no reason that interrupts would have been disabled at that point, so I probably is just to be sure.
https://github.com/z88dk/z88dk/pull/2665/commits/b23f470dc6c8c2876d00f7c34ed7ce291cc97514